docs/spec_draft.md

```diff
@@ -341,6 +351,174 @@ only exists to establish data dependencies from `result` to `inputs`.
+
+For each group `G` $\in$ `groups`, the `result`, to be visible by each member of
+`G`, is given by:
+`concatenate(operands, all_gather_dim)`, where `operands` =
+{ `operand` corresponding to the execution instances, following their order, in
+`G` }.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
+| `all_gather_dim`        | constant of type `si64`                                                                                                                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64`                                                                                                  |
+| `channel_handle`        | constant of type struct { `channel-id`: `si64`, `channel-type`: enum of `CHANNEL_TYPE_INVALID`, `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, `HOST_TO_DEVICE` } |
+| `use_global_device_ids` | constant of type `boolean`                                                                                                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `all_gather_dim` $\in$ [0, rank(`operand`)).
+  * (C2) dim(`operand`, `all_gather_dim`) $\ne$ 0.
+  * (C3) On `replica_groups`:
+    * All the subgroups in `replica_groups` have the same size.
+    * The replica ids contains all values in the range [0, N) where N is the
+      number of replicas in `replica_groups`.
+    * An empty `replica groups` is allowed only when `use_global_device_ids` is
+      `false`.
+  * (C4) dim(`result`, `all_gather_dim`) % dim(`operand`, `all_gather_dim`) $=$ 0.
+  * (C5) element_type(`result`) $=$ element_type(`operand`).
+  * (C6) rank(`result`) $=$ rank(`operand`).
+  * (C7) dim(`result`, i) is given by:
+    * dim(`operand`, i), i $\ne$ `all_gather_dim`.
+    * `shard_count` * dim(`operand`, i), otherwise, where `shard_count` $=$
+      dim(`result`, `all_gather_dim`) / dim(`operand`, `all_gather_dim`).
+
+### Examples
+
+```mlir
+// %operand (at ei(0, 0)): [[1.0, 2.0], [3.0, 4.0]]
+// %operand (at ei(1, 0)): [[5.0, 6.0], [7.0, 8.0]]
+%result = "stablehlo.all_gather"(%operand) {
+  all_gather_dim = 1 : i64,
+  replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>
+} : (tensor<2x2xf32>) -> tensor<2x4xf32>
+//
+// Assuming num_pids = 1, a single group [ei(0, 0), ei(1, 0)] is formed.
+//
+// %result (at ei(0, 0) or ei(1, 0): [[1.0, 2.0, 5.0, 6.0], [3.0, 4.0, 7.0, 8.0]]
```

Comment:
Given the feedback above and the notation proposed in #634, this would read:

```
// num_replicas: 2
// num_partitions: 1
// %operand<0, 0>: [[1.0, 2.0], [3.0, 4.0]]
// %operand<1, 0>: [[5.0, 6.0], [7.0, 8.0]]
%result = "stablehlo.all_gather"(%operand) {
  all_gather_dim = 1 : i64,
  replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>
  // use_global_device_ids = false
} : (tensor<2x2xf32>) -> tensor<2x4xf32>
// %result<0, 0> [[1.0, 2.0, 5.0, 6.0], [3.0, 4.0, 7.0, 8.0]]
// %result<1, 0> [[1.0, 2.0, 5.0, 6.0], [3.0, 4.0, 7.0, 8.0]]
```